Update python verions

robbiemu opened the pull request:
currently, and for quite some time actually, we have been in a weird situation where no python install is possible without using flags, at least with standard setups such as uv. here's an example:

```bash
uv venv --python 3.9
Using CPython 3.9.6 interpreter at: /Applications/Xcode.app/Contents/Developer/usr/bin/python3
Creating virtual environment at: .venv
Activate with: source .venv/bin/activate

source .venv/bin/activate

uv pip install -r requirements.txt
  × No solution found when resolving dependencies:
  ╰─▶ Because only the following versions of numpy are available:
          numpy<1.26.4
          numpy>1.27.dev0
      and you require numpy>=1.26.4,<1.27.dev0, we can conclude that your requirements are unsatisfiable.

      hint: `numpy` was requested with a pre-release marker (e.g., numpy>=1.26.4,<1.27.dev0), but
      pre-releases weren't enabled (try: `--prerelease=allow`)

      hint: `numpy` was found on https://download.pytorch.org/whl/cpu, but not at the requested version
      (numpy>=1.26.4,<1.27.dev0). A compatible version may be available on a subsequent index (e.g.,
      https://download.pytorch.org/whl/cpu). By default, uv will only consider versions that are published
      on the first index that contains a given package, to avoid dependency confusion attacks. If all
      indexes are equally trusted, use `--index-strategy unsafe-best-match` to consider all versions from
      all indexes, regardless of the order in which they were defined.
      
uv pip install --index-strategy unsafe-best-match -r requirements.txt
  × No solution found when resolving dependencies:
  ╰─▶ Because the current Python version (3.9.6) does not satisfy Python>=3.10 and matplotlib>=3.10.0
      depends on Python>=3.10, we can conclude that matplotlib>=3.10.0 cannot be used.
      And because only the following versions of matplotlib are available:
          matplotlib<=3.10.0
          matplotlib==3.10.1
          matplotlib==3.10.3
      and you require matplotlib>=3.10.0, we can conclude that your requirements are unsatisfiable.

      hint: `matplotlib` was requested with a pre-release marker (e.g., all of:
          matplotlib>3.10.0,<3.10.1
          matplotlib>3.10.1,<3.10.3
          matplotlib>3.10.3,<3.11.dev0
      ), but pre-releases weren't enabled (try: `--prerelease=allow`)
```

The numpy thing was an issue at least straight through torch 2.5 (torch 2.5.1 switched to numpy 2.x). Torch never published their own numpy 1.26.4, switching at 1.26.3 to 2.0, and uv stops with the first valid provider and its already on torch, so uv gives up when it finds the provider doesn't have the version it needs.

this same install works starting at 3.10, (numpy aside). But starting at python 3.13, we get a problem with torch:

```bash
uv pip install --index-strategy unsafe-best-match -r requirements.txt
  × No solution found when resolving dependencies:
  ╰─▶ Because only the following versions of torch are available:
          torch<=2.2.1
          torch==2.2.1+cpu
          torch==2.2.2
          torch==2.2.2+cpu
          torch>2.3.dev0
      and torch>=2.2.1,<=2.2.2+cpu has no wheels with a matching Python ABI tag (e.g., `cp313`), we can
      conclude that torch>=2.2.1,<=2.2.2+cpu cannot be used.
      And because you require torch>=2.2.1,<2.3.dev0, we can conclude that your requirements are
      unsatisfiable.

      hint: `torch` was requested with a pre-release marker (e.g., all of:
          torch>2.2.1,<2.2.1+cpu
          torch>2.2.1+cpu,<2.2.2
          torch>2.2.2,<2.2.2+cpu
          torch>2.2.2+cpu,<2.3.dev0
      ), but pre-releases weren't enabled (try: `--prerelease=allow`)

      hint: You require CPython 3.13 (`cp313`), but we only found wheels for `torch` (v2.2.2+cpu) with the
      following Python ABI tags: `cp38`, `cp39`, `cp310`, `cp311`, `cp312`
```

BTW, PyTorch (>= 2.3) officially [dropped support](https://github.com/pytorch/pytorch/issues/129805) for Python 3.9 as of June 22, 2024, following the NEP 29 deprecation schedule. So it really makes sense to bump this project up to 3.10.

## Can we just update Pytorch?

I see 3 files that use Pytorch in the project.

- ./convert_hf_to_gguf.py
- ./convert_lora_to_gguf.py
- ./tools/mtmd/legacy-models/minicpmv-convert-image-encoder-to-gguf.py


I asked gemini to help look for issues. It notes:

---

Okay, I have analyzed the PyTorch usage in the three provided scripts: `./convert_hf_to_gguf.py`, `./convert_lora_to_gguf.py`, and `./tools/mtmd/legacy-models/minicpmv-convert-image-encoder-to-gguf.py`.

Here's the verification for each file regarding compatibility with PyTorch 2.3 or greater:

**1. `./convert_hf_to_gguf.py`**

*   **PyTorch Imports:**
    *   `import torch`
    *   `from torch import Tensor` (for type hinting)
*   **Key PyTorch Usages & Verification:**
    *   **`torch.load(str(self.dir_model / part_name), map_location="cpu", mmap=True, weights_only=True)`**:
        *   This is the primary method for loading PyTorch model parts if they are not in `safetensors` format.
        *   Crucially, it **explicitly sets `weights_only=True`**. This is the recommended security practice and aligns with PyTorch's direction. It means that even when PyTorch 2.6+ changes the default of `weights_only` to `True`, this script's behavior will remain consistent and safe.
        *   **Compatibility:** This usage is perfectly compatible with PyTorch 2.3+ and future versions.
    *   **Tensor Attributes and Methods:**
        *   `.dtype` (e.g., `first_tensor.dtype == torch.float16`, `data_torch.dtype`)
        *   `.to(torch.float32)` (and other type conversions like to `torch.float16`, `torch.bfloat16`)
        *   `.numpy()`
        *   `.shape`
        *   `.reshape(...)`, `torch.cat(...)`, `torch.stack(...)`, `torch.permute(...)`, `torch.transpose(...)`, `torch.div_(...)` (in-place), `torch.view(...)` (used in various model-specific `modify_tensors` methods).
        *   `torch.arange(...)`, `torch.tensor(...)`, `torch.exp(...)`, `torch.equal(...)`, `torch.norm(...)`, `torch.from_numpy(...)` (via `torch.frombuffer` in `LazyTorchTensor`), `torch.split(...)`.
        *   **Compatibility:** All these are fundamental and stable PyTorch tensor operations and attributes. No breaking changes affecting these are anticipated in PyTorch 2.3+.
    *   **`LazyTorchTensor` Class:**
        *   This custom class uses `torch.empty` for meta tensor creation and `torch.frombuffer` for creating tensors from remote data.
        *   It implements `__torch_function__` to wrap PyTorch functions, which is a stable PyTorch feature.
        *   The dtypes used (`torch.float64`, `torch.float32`, `torch.bfloat16`, `torch.float16`, `torch.int64`, etc.) are all standard.
        *   **Compatibility:** The PyTorch interactions within this class are stable.
    *   **`torch.inference_mode()`:**
        *   Used as a context manager.
        *   **Compatibility:** Standard and stable.
*   **Conclusion for `./convert_hf_to_gguf.py`:**
    *   This script **is compatible** with PyTorch 2.3 or greater. The explicit use of `weights_only=True` in `torch.load` is key. All other PyTorch operations are standard and stable.

**2. `./convert_lora_to_gguf.py`**

*   **PyTorch Imports:**
    *   `import torch`
    *   `from torch import Tensor` (for type hinting)
    *   It also imports `LazyTorchTensor` and `ModelBase` from `./convert_hf_to_gguf.py`, inheriting their PyTorch usage patterns.
*   **Key PyTorch Usages & Verification:**
    *   **`torch.load(input_model, map_location="cpu", weights_only=True)`**:
        *   Used as a fallback if `adapter_model.safetensors` is not found, loading `adapter_model.bin`.
        *   It **explicitly sets `weights_only=True`**.
        *   **Compatibility:** Same as in the first script, this is perfectly compatible and future-proof regarding `weights_only` defaults.
    *   **`LoraTorchTensor` Class:**
        *   This custom class wraps LoRA A and B tensors.
        *   Its methods (`__getitem__`, `reshape`, `view`, `permute`, `transpose`, `to`, `__torch_function__` for `torch.stack`, `torch.cat`) perform operations on the underlying PyTorch tensors using standard PyTorch tensor methods. For example, `self._lora_A.reshape(...)`.
        *   Uses `lora_a.T` (tensor transpose property).
        *   **Compatibility:** All internal operations rely on stable PyTorch tensor functionalities.
    *   **Inheritance from `ModelBase`:**
        *   The `LoraModel` class inherits from `ModelBase` (defined in `convert_hf_to_gguf.py`). It reuses the tensor processing logic, which, as established above, is compatible.
        *   Uses `LazyTorchTensor.from_eager(tensor)`.
    *   **`torch.inference_mode()`:**
        *   Used as a context manager.
        *   **Compatibility:** Standard and stable.
*   **Conclusion for `./convert_lora_to_gguf.py`:**
    *   This script **is compatible** with PyTorch 2.3 or greater. It benefits from the safe `torch.load` usage and the compatible base class from `convert_hf_to_gguf.py`.

**3. `./tools/mtmd/legacy-models/minicpmv-convert-image-encoder-to-gguf.py`**

*   **PyTorch Imports:**
    *   `import torch`
    *   `import torch.nn.functional as F`
    *   `from torch import nn`
*   **Key PyTorch Usages & Verification:**
    *   **Model Definitions (e.g., `SiglipVisionEmbeddings`, `SiglipAttention`, `Idefics2VisionTransformer`):**
        *   These classes inherit from `torch.nn.Module` or `PreTrainedModel` (which itself is an `nn.Module`).
        *   They use standard PyTorch layers: `nn.Conv2d`, `nn.Embedding`, `nn.Linear`, `nn.LayerNorm`, `nn.ModuleList`.
        *   Standard activation functions are used (via `ACT2FN` or directly like `F.pad`).
        *   Weight initialization functions (`nn.init.normal_`, `nn.init.zeros_`, `_trunc_normal_`, `variance_scaling_`) use basic tensor operations like `tensor.uniform_`, `tensor.erfinv_()`, `tensor.mul_()`, `tensor.add_()`, `tensor.clamp_()`, `tensor.normal_()`.
        *   **Compatibility:** These are core, stable components of `torch.nn` and associated tensor manipulations. `tensor.erfinv_` is a standard operation.
    *   **`torch.load(os.path.join(dir_model, "minicpmv.clip"))`**:
        *   This call loads the weights for the `Idefics2VisionTransformer` (or `SiglipVisionTransformer`).
        *   **Crucially, `weights_only` is NOT specified.**
        *   In PyTorch 2.3, 2.4, and 2.5, the default for `weights_only` is `False`. This means the script will load the file as is.
        *   The loaded object is then passed to `model.load_state_dict(...)`. This function expects a dictionary mapping tensor names to `torch.Tensor` objects (a state dictionary).
        *   If `minicpmv.clip` is a pickled file containing only a state dictionary (which is highly likely given its use with `load_state_dict`), then:
            *   Loading with `weights_only=False` (default in 2.3-2.5) works fine.
            *   Loading with `weights_only=True` would also work fine and would be safer.
            *   When PyTorch 2.6+ changes the default to `weights_only=True`, this line of code will *continue to work correctly* if `minicpmv.clip` is indeed just a state dictionary of tensors.
        *   **Compatibility:** This usage is compatible with PyTorch 2.3, 2.4, and 2.5. While not an immediate breaking issue for these versions, it's recommended to add `weights_only=True` to this call if `minicpmv.clip` is confirmed to be a safe state dictionary, for future-proofing and best practice.
    *   **Tensor Manipulations:**
        *   `state_dict = model.state_dict()`
        *   `data.squeeze().numpy()`
        *   `v.chunk(3, dim=0)[0]`, `v.transpose(-1, -2).contiguous()`, `v.unsqueeze(0)` (in `_replace_name_resampler` and `_replace_name`).
        *   `torch.from_numpy(...)` (used implicitly when `torch.from_numpy(get_2d_sincos_pos_embed(...))` is called).
        *   **Compatibility:** All these are standard and stable PyTorch operations.
    *   **`torch.no_grad()`:** Used as a context manager.
        *   **Compatibility:** Standard and stable.
*   **Conclusion for `./tools/mtmd/legacy-models/minicpmv-convert-image-encoder-to-gguf.py`:**
    *   This script **is compatible** with PyTorch 2.3 or greater.
    *   The `torch.load` call without an explicit `weights_only` parameter is not a breaking issue for PyTorch 2.3-2.5 for loading state dictionaries. It will also likely work with PyTorch 2.6+ if the file is a standard state dictionary. However, for best practice and to eliminate any ambiguity, adding `weights_only=True` to this specific `torch.load` call is advisable, assuming `minicpmv.clip` contains only a state dictionary.

**Overall Conclusion:**

All three scripts (`./convert_hf_to_gguf.py`, `./convert_lora_to_gguf.py`, and `./tools/mtmd/legacy-models/minicpmv-convert-image-encoder-to-gguf.py`) **are compatible with upgrading PyTorch to 2.3 or greater.**

*   The critical `torch.load` calls in the first two scripts explicitly use `weights_only=True`, which is ideal.
*   The `torch.load` call in the third script does not specify `weights_only`, but given its context (loading a state dictionary), this is not a breaking issue for PyTorch 2.3-2.5 and is likely to remain compatible with PyTorch 2.6+ for this specific use case. Adding `weights_only=True` to it is a good recommendation for robustness.
*   All other PyTorch API usages identified are core, stable functionalities that are not subject to breaking changes that would affect these scripts in PyTorch 2.3+.

---

So, I created a script to check general availability of torch at python version: https://gist.github.com/robbiemu/c224d6daf2f05d918ac8cf759db33288

uv run main.py --toml $GH/llama.cpp/pyproject.toml --requirements $GH/llama.cpp/requirements.txt --targets cpu,cuda
Building dynamic compatibility grid — this may take a few seconds…
Stable variants discovered: cpu, cu100, cu101, cu102, cu110, cu111, cu113, cu115, cu116, cu117, cu118, cu121, cu124, cu126, cu128, cu92, rocm3, rocm4, rocm5, rocm6, xpu
Nightly variants discovered: cpu, cu118, cu126, cu128, rocm6, xpu
Python \ Torch | 2.2 | 2.3 | 2.4 | 2.5 | 2.6 | 2.7 | 2.8
---------------+-----+-----+-----+-----+-----+-----+----
3.9 | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | !
3.10 | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | !
3.11 | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | !
3.12 | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | !
3.13 | X | X | X | ✓ | ✓ | ✓ | !

To support all python versions from 3.10 forward, we will need to upgrade torch to 2.5.

Taken together, we should upgrade to python >=3.10, torch 2.5.1, and numpy >2.0.

So, I wrote a small script to crawl the repo for "import numpy" and "from numpy" and collect all the numpy members used in the files, and asked gemini to check them for safety in upgrading to numpy 2.x, and verified our usage was compatible:

Now, let's verify these against potential breaking changes, primarily focusing on the NumPy 1.x to NumPy 2.0 transition, as this is the most significant recent shift. You're aiming for `numpy>=2.0.0` to go with `torch>=2.5.0` and Python 3.10+.

**Verification of NumPy Members for NumPy 2.0 Compatibility:**

1.  **Core Array Object and Creation:**
    *   `ndarray`, `array`, `asarray`, `empty`, `ones`, `zeros`, `arange`, `frombuffer`, `fromfile`, `fromiter`: These are fundamental and remain. Their core behavior is stable.
    *   `memmap`: Stable.
    *   **Status:** ✅ Safe.

2.  **Data Types (dtypes):**
    *   `dtype` (the object itself and as a parameter): Stable.
    *   `float16`, `float32`, `float64`: Stable. These are concrete types.
    *   `int8`, `int16`, `int32`, `int64`: Stable.
    *   `uint8`, `uint16`, `uint32`, `uint64`: Stable.
    *   `bool_`: Stable (`np.bool_` is the canonical boolean type).
    *   `complex64`: Stable.
    *   `generic` (likely `np.generic` for type checking): Stable.
    *   `signedinteger` (likely `np.signedinteger` for type checking): Stable.
    *   `DTypeLike`, `NDArray` (from `numpy.typing`): These are for type hinting and are designed to work with NumPy 2.0.
    *   **Potential Issue (Minor & Unlikely Here):** The main breaking change in NumPy 2.0 regarding types was the removal of aliases like `np.int`, `np.float`, `np.bool`. Your list shows specific types (`np.int32`, `np.float64`, etc.) or `np.bool_`, which are the correct, non-deprecated forms. So, this common pitfall is **avoided**.
    *   **Status:** ✅ Safe.

3.  **Array Manipulations:**
    *   `concatenate`, `stack`, `split`, `array_split`, `hsplit`: Stable.
    *   `reshape` (implied, as `ndarray` objects have this method, though not explicitly listed as `np.reshape`): Stable.
    *   `moveaxis`, `take_along_axis`: Stable.
    *   `pad`: Stable.
    *   `packbits`, `unpackbits`: Stable.
    *   `clip`: Stable.
    *   `triu`: Stable.
    *   **Status:** ✅ Safe.

4.  **Mathematical Operations & Functions:**
    *   `cos`, `sin`, `exp`, `sqrt`, `floor`, `trunc`, `sign`: Stable.
    *   `sum`, `prod`, `argmax`, `count_nonzero`: Stable.
    *   `dot`, `einsum`: Stable.
    *   `linalg` (e.g., `np.linalg.norm`): The `np.linalg` submodule is stable.
    *   `all`, `allclose`, `array_equal`: Stable.
    *   `fmax`, `fmin`: Stable.
    *   `nan` (`np.nan`): Stable.
    *   `fft` (likely `np.fft` submodule): Stable.
    *   `hanning` (likely `np.hanning` for window functions): Stable.
    *   **Status:** ✅ Safe.

5.  **Utilities & Submodules:**
    *   `random` (e.g., `np.random.random`): The `np.random` module and its functions are stable.
    *   `lib` (e.g., `np.lib.stride_tricks`): This submodule contains utilities. While some very internal parts could change, common uses are generally stable. `np.lib.pad` is now `np.pad`. Your list shows `pad`, suggesting the direct `np.pad` is likely used, which is good.
    *   `errstate`: Stable (context manager for floating point errors).
    *   `set_printoptions`: Stable.
    *   `meshgrid`: Stable.
    *   `where`: Stable.
    *   **Status:** ✅ Safe.

**Overall Assessment for NumPy >= 2.0.0:**

Based on the list of members your script identified:

*   **The vast majority of these NumPy features are fundamental and have stable APIs that are fully compatible with NumPy 2.0.**
*   The most common breaking change in NumPy 2.0 (removal of `np.int`, `np.float` aliases) does not appear to be an issue here, as the list shows specific dtypes (e.g., `np.float32`, `np.int64`).
*   The type promotion rules in NumPy 2.0 are more consistent. This is generally a good thing and unlikely to break code unless it relied on very specific, possibly ambiguous, old promotion behaviors. For the types of operations typically done in these scripts (data conversion, basic math on well-defined types), this is usually not a problem.

**Conclusion:**

Given the specific NumPy members used by the `llama.cpp` Python scripts, an upgrade to **`numpy>=2.0.0` is highly likely to be safe and successful.**

The identified usages are predominantly core NumPy functionalities that have remained stable or had very clear migration paths that an active project would typically handle. The benefits of moving to NumPy 2.0 (better Python 3.13+ support, API consistency, potential performance improvements) outweigh the minimal risks for this particular set of functions.

This confirms the earlier recommendation to set your NumPy requirement to `numpy>=2.0.0` when using `torch>=2.5.0` and targeting Python 3.10+.

After updating the requirements (for torch 2.5.1 and torchvision) and pyproject.toml, and installing like:

```bash
rm -rf .venv

uv venv --python 3.10
Using CPython 3.10.17
Creating virtual environment at: .venv
Activate with: source .venv/bin/activate

uv pip install --index-url https://pypi.org/simple --index-strategy unsafe-best-match --extra-index-url https://download.pytorch.org/whl/ -r requirements.txt
Resolved 73 packages in 780ms
Prepared 2 packages in 1.50s
Installed 73 packages in 578ms
...
 + numpy==1.26.4
...
```
and we can verify that numpy v2 would always be acceptable:
```bash
pipdeptree --reverse --packages numpy --warn silence

numpy==1.26.4
├── contourpy==1.3.2 [requires: numpy>=1.23]
│   └── matplotlib==3.10.3 [requires: contourpy>=1.0.1]
│       └── seaborn==0.13.2 [requires: matplotlib>=3.4,!=3.6.1]
├── pandas==2.2.3 [requires: numpy>=1.22.4]
│   └── seaborn==0.13.2 [requires: pandas>=1.2]
├── seaborn==0.13.2 [requires: numpy>=1.20,!=1.24.0]
├── gguf==0.16.3 [requires: numpy>=1.17]
├── matplotlib==3.10.3 [requires: numpy>=1.23]
│   └── seaborn==0.13.2 [requires: matplotlib>=3.4,!=3.6.1]
├── torchvision==0.20.1 [requires: numpy]
└── transformers==4.46.3 [requires: numpy>=1.17]
```

and torch 2.5.1 was compiled on numpy 2.x when numpy 2.1 was current, per this [nvidia thread](https://forums.developer.nvidia.com/t/issues-with-new-pytorch-version-for-jp-6-1/309079/7), so >=2.1 makes sense. 

This finally uncovers another dependency issue stemming from Pytorch's own published version of requests (they have literally only one):
```bash
uv pip install -r requirements/requirements-all.txt
  × No solution found when resolving dependencies:
  ╰─▶ Because only requests==2.28.1 is available and you require requests>=2.32.3, we can conclude that
      your requirements are unsatisfiable.

      hint: `requests` was found on https://download.pytorch.org/whl/cpu, but not at the requested
      version (requests>=2.32.3). A compatible version may be available on a subsequent index (e.g.,
      https://download.pytorch.org/whl/cpu). By default, uv will only consider versions that are published
      on the first index that contains a given package, to avoid dependency confusion attacks. If all
      indexes are equally trusted, use `--index-strategy unsafe-best-match` to consider all versions from
      all indexes, regardless of the order in which they were defined.
```

and this is easy, because they are both minor versions is a _very_ stable library, and the version numbers are actually close. we can reasonably set requests>=2.28.1. with that, we get at the easy install you'd expect:

```bash
uv pip install -r requirements/requirements-all.txt
Resolved 84 packages in 949ms
Prepared 17 packages in 2.45s
Installed 84 packages in 509ms
...
```

And if we want the latest, we can use `uv pip install --upgrade -r requirements/requirements-all.txt --resolution=highest --index-strategy unsafe-best-match`

robbiemu committed:
Updates Python and torch versions

Updates the allowed Python version to include 3.10 up to, but not including, 3.14.

Sets the minimum torch version to 2.5 in multiple dependency files.

robbiemu committed:
Updates torchvision dependency

Updates torchvision dependency to a more recent version.

This ensures compatibility with the current torch version
and leverages the latest features and bug fixes.

robbiemu committed:
Updates NumPy dependency to version 2.1 or higher

Increases the minimum required NumPy version to 2.1 across multiple project dependencies. This ensures compatibility with newer features and improvements in the NumPy library.

Updates also include a minor patch to Torch version.

robbiemu commented:
Were those checks already not passing? I can look into them/ changes required

esrakorkmz commented:
> Were those checks already not passing? I can look into them/ changes required

Thank you. I am not really sure why I am low-coding. I don't understand much.

robbiemu committed:
pyright-based changes for  tools/mtmd/legacy-models/convert_image_encoder_to_gguf.py

Fix torch 2.5.1 / numpy 2.x compatibility in convert_image_encoder_to_gguf.py

- Updated Tensor-to-array conversions to use `np.asarray(..., dtype=...)` per NumPy 2.x migration rules (avoids copy error on float16).
- Used explicit typing and `cast(...)` to guide Pyright/Pylance under torch 2.5.1:
  - Annotated `model` as PreTrainedModel.
  - Re-cast `model.vision_model` to `CLIPVisionTransformer` to safely access `.encoder.layers`.
  - Replaced slice assignment with `__init__` to reset ModuleList contents.
- Verified compatibility by converting `openai/clip-vit-base-patch32` using `--clip-model-is-openclip`.

robbiemu committed:
Fixes type annotation for numpy arrays

Corrects type annotations for numpy arrays to allow for a broader range of numpy dtypes and resolves type checking errors.

Removes NumPy DTypeLike type hint

Updates type hints for NumPy compatibility by removing DTypeLike.

Ensures alignment with NumPy's typing system, preventing potential
type-related issues.

robbiemu committed:
Updates type hints for numpy compatibility

Updates type hints related to numpy to use `np.dtype[Any]` instead of `DTypeLike` for better compatibility and clarity. This resolves potential issues with type checking and ensures accurate type annotations for numpy-related operations within the `gguf` library.

robbiemu committed:
Better fix than earlier for quants.py

after researching the typing issues introduced in numpy 2.2

robbiemu committed:
Several types in numpy 2.2+ are defined as typeAliases with no underlying class (such as np.float32)

looking at https://github.com/microsoft/pyright/issues/9051, they declined to fix it themselves, and suggested instead that the used must add a # pyright: ignore or # type: ignore directive to suppress this error.

Numpy is working to resolve them: https://github.com/numpy/numpy/issues/28076 and has already done so with npfloat64 (which I can verify in our errors) -- see https://github.com/numpy/numpy/issues/27957 .

compilade commented on tools/mtmd/legacy-models/convert_image_encoder_to_gguf.py:
```diff
@@ -6,6 +6,10 @@
 from gguf import *
+from typing import cast
+from torch.nn import ModuleList
+from transformers.models.clip.modeling_clip import CLIPVisionTransformer
```
I think it could be reasonable to add `"reportPrivateImportUsage": "none",` in `pyrightconfig.json`. Otherwise there's also some errors in other scripts regarding `AutoConfig` and `AutoTokenizer`.

robbiemu replied:
should I remove changes to */legacy_models/* from this PR?

compilade commented on gguf-py/gguf/gguf_reader.py:
```diff
@@ -104,7 +106,7 @@ class ReaderTensor(NamedTuple):
     n_bytes: int
     data_offset: int
-    data: npt.NDArray[Any]
+    data: np.ndarray
```
```suggestion
    data: np.ndarray[Any, Any]
```

Keep the annotations where possible when changing `npt.NDArray` to `np.ndarray`.

robbiemu replied:
actually given the type fix I added to this file, I should have reverted this change entirely. Depending on final say from you, I'll either revert it or do as you say here when I roll back the pyright hint at the top.

robbiemu commented on gguf-py/gguf/gguf_reader.py:
```diff
@@ -1,3 +1,4 @@
+# pyright: reportInvalidTypeForm=false
```
we'll I am happy to do that but let me first just verify with you. The GitHub action for pyright _already_ is using python 3.11: https://github.com/ggml-org/llama.cpp/blob/6aa892ec2aa7fe0c93e87c4b970d83a942fb9454/.github/workflows/python-type-check.yml#L31

also, this is not a numpy 2.x issue, this is the more recent 2.2 issue. Several types in numpy 2.2+ are defined as typeAliases with no underlying class (such as np.float32). looking at https://github.com/microsoft/pyright/issues/9051, pyright declined to fix it themselves, and suggested instead that the used must add a # pyright: ignore or # type: ignore directive to suppress this error.

Numpy is working to resolve them: https://github.com/numpy/numpy/issues/28076 and has already done so with npfloat64 (which I can verify in our errors) -- see https://github.com/numpy/numpy/issues/27957 .

compilade replied:
> The GitHub action for pyright already seems it is using python 3.11:

Sure, but Pyright is not using 3.11 features when type-checking here since that is controlled by the `pyrightconfig.json` file which currently has `3.9` as the `pythonVersion`.

https://github.com/ggml-org/llama.cpp/blob/1dfbf2cf3a9f15193dd893396d07762bbd2c4785/pyrightconfig.json#L3

robbiemu replied:
that was a miss on my part, I should have updated the config to use 3.10 -- applying this change, and I'll roll back one of the comments to see if those are still needed.

robbiemu replied:
ok, so I have done this change and selected quants.py as a test file to see if just updating the python version is sufficient, but it is not:

  "type[type]" is not assignable to "type[dtype[Any]]" (reportArgumentType)
/home/runner/work/llama.cpp/llama.cpp/gguf-py/gguf/quants.py:171:79 - error: Argument of type "float32" cannot be assigned to parameter "otype" of type "dtype[Any]" in function "_apply_over_grouped_rows"
  "type[type]" is not assignable to "type[dtype[Any]]" (reportArgumentType)
Error: Argument of type "float32" cannot be assigned to parameter "otype" of type "dtype[Any]" in function "_apply_over_grouped_rows

I think your preference is to add the setting to pyrightconfig.json rather than adding comments to the top of the individual files. Do I have that correct?

robbiemu committed:
revert change from NDArray in this file

robbiemu committed:
Updates pyright configuration

Updates the pyright configuration to use Python 3.10.

This ensures compatibility with newer language features
and libraries. It also excludes the 'tools/legacy' and 'tests'
directories from analysis.

robbiemu committed:
Revert tools/mtmd/ to match master

robbiemu committed:
Revert "Revert tools/mtmd/ to match master"

This reverts commit eb6499cc7852c4c69e8628863d2b548d837a235f.

robbiemu commented:
I made a mistake in rolling back the changes to tools/mtmd (to protect the legacy files), and now the git history of my branch will blame this pr in the history of changes to some files which it should not. for this reason, I am going to recreate this pr. I am sorry for the error.

compilade commented:
> I made a mistake in rolling back the changes to tools/mtmd (to protect the legacy files), and now the git history of my branch will blame this pr in the history of changes to some files which it should not. for this reason, I am going to recreate this pr. I am sorry for the error.

The commits are squashed when merging to master anyway, so if the files are not changed in `git diff --merge-base master`, this will probably not be a problem.

robbiemu committed:
reverting change to legacy-models script

compilade commented:
> I made a mistake in rolling back the changes to tools/mtmd (to protect the legacy files), and now the git history of my branch will blame this pr in the history of changes to some files which it should not. for this reason, I am going to recreate this pr. I am sorry for the error.

The commits are squashed when merging to master anyway, so if the files are not changed in `git diff --merge-base master`, this will probably not be a problem.

And if you manage to fix the history of the branch, you can also force-push that to the branch of this PR and GitHub will only show the commits belonging to the new tip of the branch.

robbiemu committed:
Excludes legacy models from type checking

Updates the pyright configuration to exclude the legacy models directory from type checking.

robbiemu committed:
Configures pyright to ignore invalid type forms

Temporarily disables `reportInvalidTypeForm` in pyright configuration due to ongoing transition to the new type system in numpy 2.2.x.

compilade commented on gguf-py/gguf/quants.py:
```diff
@@ -26,7 +24,7 @@ def quant_shape_from_byte_shape(shape: Sequence[int], quant_type: GGMLQuantizati
 
 # This is faster than np.vectorize and np.apply_along_axis because it works on more than one row at a time
-def _apply_over_grouped_rows(func: Callable[[np.ndarray], np.ndarray], arr: np.ndarray, otype: DTypeLike, oshape: tuple[int, ...]) -> np.ndarray:
+def _apply_over_grouped_rows(func: Callable[[np.ndarray], np.ndarray], arr: np.ndarray, otype: np.dtype[Any], oshape: tuple[int, ...]) -> np.ndarray:
```
Using `DTypeLike` here (and in the other places where it was previously used) would likely fix some of the errors (since it seems like e.g. `np.uint32` is not an instance of `np.dtype[Any]`, but is `DTypeLike`.).

robbiemu replied:
Actually look at the latest pyright tests. this is working, replacing DTypeLike was never an issue

compilade commented on pyrightconfig.json:
```diff
@@ -1,7 +1,8 @@
-  "pythonVersion": "3.9",
+  "pythonVersion": "3.10",
   "pythonPlatform": "All",
+  "reportInvalidTypeForm": false, // TODO: remove once numpy 2.2.x resolves the transition to their new type system
```
I don't think this is necessary with the `pythonVersion` as `3.10`; I did not see this error kind in the checks of the previous commit. (But I may be mistaken)

robbiemu replied:
I think it actually is not working, as the comment at the top in each file was. I am kind of surprised by this but I see:

https://github.com/ggml-org/llama.cpp/actions/runs/15123815358/job/42511921578?pr=13574

(compare to https://github.com/ggml-org/llama.cpp/actions/runs/15122601593 when we had comments at the top of each file and testing in 3.8)

compilade replied:
I still think setting `reportInvalidTypeForm` to `false` is not necessary (since the version change to `3.10` in `pyrightconfig.json`) and will hide future problems.

> In both NumPy 1.x and 2.2.x, np.uint32 and np.dtype(np.uint32) are functionally equivalent when used as parameters in array creation or type specification

Yes, since both are `DTypeLike`. Why not simply directly use `DTypeLike` in the type annotations as before?

robbiemu replied:
they were removed because on Friday, I believe numpy 2.2.5 was current, and DtypeLike was raising errors in pyright. Specifically, I was seeing:

Error: Variable not allowed in type expression (reportInvalidTypeForm)
/home/runner/work/llama.cpp/llama.cpp/gguf-py/gguf/quants.py:29:96 - error: Variable not allowed in type expression (reportInvalidTypeForm)
Error: Variable not allowed in type expression (reportInvalidTypeForm)
/home/runner/work/llama.cpp/llama.cpp/gguf-py/gguf/quants.py:83:36 - error: Variable not allowed in type expression (reportInvalidTypeForm)
Error: Variable not allowed in type expression (reportInvalidTypeForm)

in the file before I had touched it at all.

compilade replied:
> And, I am wondering, would you like me to add it to the exclude list in pyrightconfig, since it is marked as legacy?

@robbiemu 

Without ignoring the file, it's still a relatively simple fix:

```diff
diff --git a/examples/convert_legacy_llama.py b/examples/convert_legacy_llama.py
index c4ec5c524..710cca96b 100755
--- a/examples/convert_legacy_llama.py
+++ b/examples/convert_legacy_llama.py
@@ -1308,7 +1308,7 @@ def do_dump_model(model_plus: ModelPlus) -> None:
 
 def main(args_in: list[str] | None = None) -> None:
     output_choices = ["f32", "f16"]
-    if np.uint32(1) == np.uint32(1).newbyteorder("<"):
+    if np.uint32(1) == np.uint32(1).view(np.dtype(np.uint32).newbyteorder("<")):
         # We currently only support Q8_0 output on little endian systems.
         output_choices.append("q8_0")
     parser = argparse.ArgumentParser(description="Convert a LLaMA model to a GGML compatible file")
```

robbiemu replied:
Alright, that works for now :)

Starting with python 3.13, Unpickler changes:

The persistent_load method overrides the Unpickler class in an incompatible way: the base method takes 1 positional parameter, but the override takes 2. (Pylance reportIncompatibleMethodOverride)

That's not effecting our tests though :)

compilade replied:
@robbiemu but ignoring `reportInvalidTypeForm` really is **not** needed anymore, and should be removed. At least in local tests it doesn't seem to be required to ignore anymore.

robbiemu replied:
thought I had already done it! its done now.

robbiemu committed:
Uses `np.dtype` for NumPy type hints in GGUF reader

Replaces uses of bare NumPy type hints (e.g., `np.uint32`) with
`np.dtype(np.uint32)` in the GGUF reader. This aligns with changes in
NumPy 2.0+ and ensures forward compatibility.

Also applies this change to the quantize/dequantize functions in the
quants module.

robbiemu committed:
reverting removal of DTypeLike changes

DTypeLike was removed because pyright was choking on it against numpy 2.2.1 Now numpy at 2.2.6 seems to have changes needed for pyright to not have issues with it. additionally, DTypeLike is accurate in a way that np.dtype[Any] was not.

robbiemu committed:
Fixes byteorder check and removes unrequested exclusion tests

Updates the byteorder check in the conversion script to be more robust.

Removes the "tests" directory from the pyright exclude list.

compilade commented on gguf-py/gguf/gguf_reader.py:
```diff
@@ -134,12 +134,12 @@ def __init__(self, path: os.PathLike[str] | str, mode: Literal['r', 'r+', 'c'] =
 
         # Check for GGUF magic
-        if self._get(offs, np.uint32, override_order = '<')[0] != GGUF_MAGIC:
+        if self._get(offs, np.dtype(np.uint32), override_order = '<')[0] != GGUF_MAGIC:
```
Since <https://github.com/ggml-org/llama.cpp/pull/13574/commits/a5c74dd2d7284a2c9eae0b6ccf92aadd5bad2b50>, is <https://github.com/ggml-org/llama.cpp/pull/13574/commits/d49b004f8fa17e34c2df38cddec1f61a0460caeb> really needed anymore?

robbiemu replied:
I think so. You can see yesterday here: d49b004f8fa17e34c2df38cddec1f61a0460caeb I went back and re-added these after removing the type ignore hint, because it was not passing: https://github.com/ggml-org/llama.cpp/actions/runs/15123815358

however, yes, I _do not_ need it for that specific line (and another like it). however, if I undo them all:

```
git checkout 6ee825381b5cdecedb42d4ca5b43d1cf4d09f2c1 -- gguf-py/gguf/gguf_reader.py

pyright gguf-py/gguf/gguf_reader.py
/Users/Shared/Public/Github/llama.cpp/gguf-py/gguf/gguf_reader.py
  /Users/Shared/Public/Github/llama.cpp/gguf-py/gguf/gguf_reader.py:137:28 - error: Argument of type "uint32" cannot be assigned to parameter "dtype" of type "dtype[Any]" in function "_get"
    Type "uint32" is not assignable to type "dtype[Any]" (reportArgumentType)
  /Users/Shared/Public/Github/llama.cpp/gguf-py/gguf/gguf_reader.py:142:40 - error: Argument of type "uint32" cannot be assigned to parameter "dtype" of type "dtype[Any]" in function "_get"
    Type "uint32" is not assignable to type "dtype[Any]" (reportArgumentType)
  /Users/Shared/Public/Github/llama.cpp/gguf-py/gguf/gguf_reader.py:165:39 - error: Argument of type "uint64" cannot be assigned to parameter "dtype" of type "dtype[Any]" in function "_get"
    Type "uint64" is not assignable to type "dtype[Any]" (reportArgumentType)
  /Users/Shared/Public/Github/llama.cpp/gguf-py/gguf/gguf_reader.py:215:34 - error: Argument of type "uint64" cannot be assigned to parameter "dtype" of type "dtype[Any]" in function "_get"
    Type "uint64" is not assignable to type "dtype[Any]" (reportArgumentType)
  /Users/Shared/Public/Github/llama.cpp/gguf-py/gguf/gguf_reader.py:216:44 - error: Argument of type "uint8" cannot be assigned to parameter "dtype" of type "dtype[Any]" in function "_get"
    Type "uint8" is not assignable to type "dtype[Any]" (reportArgumentType)
  /Users/Shared/Public/Github/llama.cpp/gguf-py/gguf/gguf_reader.py:237:41 - error: Argument of type "uint32" cannot be assigned to parameter "dtype" of type "dtype[Any]" in function "_get"
    Type "uint32" is not assignable to type "dtype[Any]" (reportArgumentType)
  /Users/Shared/Public/Github/llama.cpp/gguf-py/gguf/gguf_reader.py:239:36 - error: Argument of type "uint64" cannot be assigned to parameter "dtype" of type "dtype[Any]" in function "_get"
    Type "uint64" is not assignable to type "dtype[Any]" (reportArgumentType)
  /Users/Shared/Public/Github/llama.cpp/gguf-py/gguf/gguf_reader.py:264:34 - error: Argument of type "uint32" cannot be assigned to parameter "dtype" of type "dtype[Any]" in function "_get"
    Type "uint32" is not assignable to type "dtype[Any]" (reportArgumentType)
  /Users/Shared/Public/Github/llama.cpp/gguf-py/gguf/gguf_reader.py:268:32 - error: Argument of type "uint64" cannot be assigned to parameter "dtype" of type "dtype[Any]" in function "_get"
    Type "uint64" is not assignable to type "dtype[Any]" (reportArgumentType)
  /Users/Shared/Public/Github/llama.cpp/gguf-py/gguf/gguf_reader.py:272:37 - error: Argument of type "uint32" cannot be assigned to parameter "dtype" of type "dtype[Any]" in function "_get"
    Type "uint32" is not assignable to type "dtype[Any]" (reportArgumentType)
  /Users/Shared/Public/Github/llama.cpp/gguf-py/gguf/gguf_reader.py:276:41 - error: Argument of type "uint64" cannot be assigned to parameter "dtype" of type "dtype[Any]" in function "_get"
    Type "uint64" is not assignable to type "dtype[Any]" (reportArgumentType)
  /Users/Shared/Public/Github/llama.cpp/gguf-py/gguf/gguf_reader.py:291:43 - error: Argument of type "uint32" cannot be assigned to parameter "dtype" of type "dtype[Any]" in function "_get"
    Type "uint32" is not assignable to type "dtype[Any]" (reportArgumentType)
  /Users/Shared/Public/Github/llama.cpp/gguf-py/gguf/gguf_reader.py:355:29 - error: Type "uint8" is not assignable to declared type "dtype[Any]"
    Type "uint8" is not assignable to type "dtype[Any]" (reportAssignmentType)
13 errors, 0 warnings, 0 informations 
```

I recommend to keep it for consistency

robbiemu committed:
reverting unnecessary changes in gguf-py/gguf as of numpy 2.2.6

robbiemu committed:
pyrightconfig.json

I thought I had already done this!

robbiemu commented:
@compilade hey I notice there is now a conflict. this was sitting ready for approval for a few days. before I fix it, can I get a sense of whether this will be approved?
